### ISSUE TITLE
CS0165 false positive

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#47642]
If this is a known limitation, consider this as a feature request instead of being a false positive report.
This is related to https://github.com/dotnet/csharplang/issues/3882

**Version Used**: 

master

**Steps to Reproduce**:

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABADAAmwEYAWAbgFgAoa4gNkJPwGEAbAewDs4BBTlZgAsInAOZwAMtBgAKCQEtYAHmK4AfPijsArggDGcAJTUA3tXwX8C5ao3z4YKPgC8+bgHcrimCqLqZWroGxlSW+GahYZYA2vZwjgB0zDqcMPgAtPhEALou+ABEAGLyCLAAklBCIuIo+fgA9PVRiAjsCCwAyrhEdACsIPgAqlBw+OwAZvjanNBQ8qLcKPgcehCs+ABuEAjyEMCsowDkcY6H5pYAvpQ0VBdAA)

```csharp
using System.Collections.Generic;

static void CloneAndChangeLast(List<string> source)
{
    List<string> items = new List<string>(source)
    {
        [items.Count - 1] = "LastIsChanged" //     error CS0165: Use of unassigned local variable 'items'
    };

}
```

**Expected Behavior**:

Code compiles to the following (just parallel behavior to when using `0` - see [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABADAAmwEYAWAbgFgAoa4gNkJPwGEAbAewDs4BBTlZgAsInAOZwAMtBgAKCQEtYAHmK4AfPijsArggDGcAJTUA3tXwX8C5ao3z4YKPgC8+bgHcrimCqLqZWroGxlSW+GahYZYA2rgAui74AEQAYvIIsACSUEIi4ihJ+AD0RVGICOwILADKuER0AKwg+ACqUHD47ABm+Nqc0FDyotwo+Bx6EKz4AG4QCPIQwKwdAOT2cI4r5pYAvpQ0VDtAA)):

```csharp
List<string> list = new List<string>(source);
list[list.Count - 1] = "FirstIsChanged";
List<string> list2 = list;
```

**Actual Behavior**:

error CS0165: Use of unassigned local variable 'items'

**Comment by svick** [comment]
This is the correct behavior, you can only reference `items` after it has been assigned and that only happens after the whole expression on the right side of `=` has been evaluated. And I don't think it would be a good idea to somehow assign partially constructed objects.

**Comment by Youssef1313** [comment]
@svick A code like the following:

```csharp
using System.Collections.Generic;

static void CloneAndChangeLast(List<string> source)
{
    List<string> items = new List<string>(source)
    {
        [0] = "LastIsChanged" //     error CS0165: Use of unassigned local variable 'items'
    };

}
```

Compile to the following based on SharpLab:

```csharp
internal static class <Program>$
{
    private static void <Main>$(string[] args)
    {
    }

    internal static void <<Main>$>g__CloneAndChangeLast|0_0(List<string> source)
    {
        List<string> list = new List<string>(source);
        list[0] = "FirstIsChanged";
        List<string> list2 = list;
    }
}
```

Note that `list[0] = "FirstIsChanged";` is done after initializing the list.

So I can't see how the list can be initialized at that point?

**Comment by RikkiGibson** [comment]
In this specific case I would be tempted to just read the length of 'source', assuming it will be the same as the final list.

**Comment by Youssef1313** [comment]
@RikkiGibson What if the signature of `CloneAndChangeLast` was `CloneAndChangeFirst(IEnumerable<string> source)` instead of `CloneAndChangeFirst(List<string> source)`?

I'll have to call the `Count` extension method which I think is less performant?

**Comment by svick** [comment]
@Youssef1313 I meant that you could observe the list in the intermediate state after the constructor has finished, but before initializers have executed. That's not as bad as being able to observe it while the constructor is still running, but I think it's still not okay.

**Comment by Youssef1313** [comment]
@svick Got your point. Would this be a blocker for https://github.com/dotnet/csharplang/issues/3882 *if* in someday it was decided to implement it?

**Comment by svick** [comment]
@Youssef1313 I don't think it would. In https://github.com/dotnet/csharplang/issues/3882, the generated code would directly access the collection that's being constructed, not the variable where it ultimately ends up. That's the crucial difference here.

**Comment by Youssef1313** [comment]
@svick Great.

I'll let @RikkiGibson give this issue a final look, and close it if no action should be taken (which most likely will be the case).

**Comment by jaredpar** [comment]
This behavior is "By Design". The local is not assigned until the initiliazer has completed.
